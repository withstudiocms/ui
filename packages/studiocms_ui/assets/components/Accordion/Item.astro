---
import Icon from '../Icon/Icon.astro';

interface Props {
	open?: boolean;
}

const { open = false } = Astro.props;
---

<div class="sui-accordion-item" class:list={{ active: open }} data-open={open}>
    <div class="sui-accordion-summary not-content" tabindex="0">
        <Icon name="chevron-right-20-solid" class='sui-summary-chevron' width={20} height={20} />
        <div class="sui-accordion-summary-content">
            <slot name="summary" />
        </div>
    </div>
    <div class="sui-accordion-details" class:list={{ active: open, open: open, initial: !open }}>
        <slot name="default" />
    </div>
</div>
