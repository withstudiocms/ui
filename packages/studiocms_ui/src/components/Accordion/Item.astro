---
import Icon from '../Icon/Icon.astro';

interface Props {
	open?: boolean;
}

const { open = false } = Astro.props;
---

<div class="sui-accordion-item" data-open={open}>
    <button class="sui-accordion-summary not-content" type="button" aria-expanded={open}>
        <Icon name="heroicons:chevron-right-20-solid" class='sui-summary-chevron' width={20} height={20} />
        <div class="sui-accordion-summary-content">
            <slot name="summary" />
        </div>
    </button>
    <div class="sui-accordion-details" aria-hidden={!open}>
        <div class="sui-accordion-content">
            <slot name="default" />
        </div>
    </div>
</div>
