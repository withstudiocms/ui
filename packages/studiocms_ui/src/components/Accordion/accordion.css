.sui-accordion {
  width: 100%;
}

.sui-accordion-item, .sui-accordion-details {
  margin-top: 0 !important;
}

.sui-accordion-details {
  margin-left: calc(2.25rem);
}

.sui-accordion-details.initial {
  visibility: hidden;
  transition: none;
  max-height: none;
  position: absolute;
  top: -200vh;
  left: -200vh;
}

.sui-accordion-details.active {
  overflow: hidden;
  max-height: 0;
  transition: all .3s ease;
  position: relative;
  bottom: .25rem;
}

.sui-accordion-details.open {
  max-height: none;
}

.sui-accordion.outlined, .sui-accordion.filled {
  border: 1px solid hsl(var(--border));
  border-radius: var(--radius-sm);
}

.sui-accordion.filled {
  background-color: hsl(var(--background-step-1));
}

.sui-accordion.separated .sui-accordion-item, 
.sui-accordion.outlined .sui-accordion-item, 
.sui-accordion.filled .sui-accordion-item {
  border-bottom: 1px solid hsl(var(--border));
}

.sui-accordion.separated .sui-accordion-item:first-of-type {
  border-top: 1px solid hsl(var(--border));
}

.sui-accordion.filled .sui-accordion-item:last-of-type,
.sui-accordion.outlined .sui-accordion-item:last-of-type {
  border-bottom: none;
}

.sui-accordion-summary {
  color: hsl(var(--text-normal));
  display: flex;
  flex-direction: row;
  gap: .5rem;
  align-items: center;
  padding-left: .5rem;
  padding-bottom: .5rem;
  padding-top: .5rem;
}

.sui-summary-chevron {
  transition: all .3s ease;
}

.sui-accordion-item.active .sui-summary-chevron {
  transform: rotate(90deg);
}
