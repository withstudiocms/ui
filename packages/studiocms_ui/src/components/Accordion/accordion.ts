const allAccordions = document.querySelectorAll<HTMLDivElement>('.sui-accordion');

for (const accordion of allAccordions) {
	const multipleOpen = accordion.dataset.multiple === 'true';

	const items = accordion.querySelectorAll<HTMLDivElement>('.sui-accordion-item');

	for (const item of items) {
		const summary = item.querySelector<HTMLDivElement>('.sui-accordion-summary');
		const content = item.querySelector<HTMLDivElement>('.sui-accordion-details');

		if (!content || !summary) continue;

		const itemWidth = item.getBoundingClientRect().width;
		content.style.width = `${itemWidth - 36}px`;

		let contentBoundingBox = content.getBoundingClientRect();

		content.style.width = 'auto';

		if (content.classList.contains('initial')) {
			content.classList.remove('initial');
			content.classList.add('active');
		}

		const open = item.dataset.open === 'true';
		if (open) {
			item.classList.add('active');
			content.style.maxHeight = `${contentBoundingBox.height}px`;
		}

		const toggleElement = () => {
			if (!multipleOpen) {
				for (const otherItem of items) {
					if (otherItem !== item) {
						otherItem.classList.remove('active');

						const otherContent = otherItem.querySelector<HTMLDivElement>('.sui-accordion-details');
						if (otherContent) {
							otherContent.style.maxHeight = '0';
						}
					}
				}
			}

			item.classList.toggle('active');

			if (item.classList.contains('active')) {
				content.style.maxHeight = `${contentBoundingBox.height}px`;
			} else {
				content.style.maxHeight = '0';
				content.classList.remove('open');
			}
		};

		summary.addEventListener('click', toggleElement);
		summary.addEventListener('keydown', (event) => {
			if (event.key === 'Enter' || event.key === ' ') {
				toggleElement();
			}
		});

		let resetting = false;

		window.addEventListener('resize', async () => {
			if (item.classList.contains('active')) {
				content.style.maxHeight = 'none';
				contentBoundingBox = content.getBoundingClientRect();
				content.style.maxHeight = `${contentBoundingBox.height}px`;
			} else if (!resetting) {
				resetting = true;

				content.classList.remove('active');
				content.classList.add('initial');

				const itemWidth = item.getBoundingClientRect().width;

				content.style.width = `${itemWidth - 36}px`;
				content.style.maxHeight = 'none';

				contentBoundingBox = content.getBoundingClientRect();

				content.style.width = 'auto';
				content.style.maxHeight = '0';

				content.classList.add('active');
				content.classList.remove('initial');

				resetting = false;
			}
		});
	}
}
