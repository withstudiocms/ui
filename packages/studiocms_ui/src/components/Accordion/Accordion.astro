---
import './accordion.css';

interface Props {
  /**
   * Whether multiple items in the accordion can be open at once. Default is `true`.
   */
  multipleOpen?: boolean;
  /**
   * The style of the accordion. Default is `outlined`
   */
  style?: 'outlined' | 'separated' | 'filled' | 'blank';
}

const { style = 'outlined', multipleOpen = true } = Astro.props;
---

<div class="sui-accordion" class:list={[style]} data-multiple={multipleOpen}>
  <slot />
</div>
<script>
  import 'studiocms:ui/scripts/accordion';
</script>
