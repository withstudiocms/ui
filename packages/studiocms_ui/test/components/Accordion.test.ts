import { describe, expect } from 'vitest';
import Accordion from '../../src/components/Accordion/Accordion.astro';
import AccordionItem from '../../src/components/Accordion/Item.astro';
import { test } from '../fixtures/vitest/AstroContainer';

describe('Accordion Component', () => {
	test('renders Accordion component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Accordion, 'Accordion');
		expect(result).toMatchSnapshot();
	});

	test('renders AccordionItem component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(AccordionItem, 'AccordionItem');
		expect(result).toMatchSnapshot();
	});

	test('renders Accordion with multiple AccordionItem components', async ({ renderComponent }) => {
		const AccordionItems: string[] = [];

		for (let i = 0; i < 3; i++) {
			const itemResult = await renderComponent(AccordionItem, `AccordionItem${i + 1}`, {
				slots: {
					default: `Content for item ${i + 1}`,
				},
			});
			AccordionItems.push(itemResult);
		}

		const result = await renderComponent(Accordion, 'Accordion', {
			slots: { default: AccordionItems.join('\n') },
		});
		expect(result).toMatchSnapshot();
	});
});
