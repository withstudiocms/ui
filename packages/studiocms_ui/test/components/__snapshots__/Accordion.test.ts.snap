// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion Component > renders Accordion component correctly 1`] = `"<div class="sui-accordion outlined" data-multiple="true">  </div> <script type="module" src="/mock/path/Accordion.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Accordion Component > renders Accordion with multiple AccordionItem components 1`] = `
"<div class="sui-accordion outlined" data-multiple="true"> <div class="sui-accordion-item" data-open="false"> <button class="sui-accordion-summary not-content" type="button" aria-expanded="false"> <svg style="min-width: 20px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="sui-summary-chevron" height="20" width="20" viewbox="0 0 20 20"><path fill="currentColor" fill-rule="evenodd" d="M8.22 5.22a.75.75 0 0 1 1.06 0l4.25 4.25a.75.75 0 0 1 0 1.06l-4.25 4.25a.75.75 0 0 1-1.06-1.06L11.94 10L8.22 6.28a.75.75 0 0 1 0-1.06" clip-rule="evenodd"/></svg> <div class="sui-accordion-summary-content">  </div> </button> <div class="sui-accordion-details" aria-hidden="true"> <div class="sui-accordion-content"> Content for item 1 </div> </div> </div>
<div class="sui-accordion-item" data-open="false"> <button class="sui-accordion-summary not-content" type="button" aria-expanded="false"> <svg style="min-width: 20px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="sui-summary-chevron" height="20" width="20" viewbox="0 0 20 20"><path fill="currentColor" fill-rule="evenodd" d="M8.22 5.22a.75.75 0 0 1 1.06 0l4.25 4.25a.75.75 0 0 1 0 1.06l-4.25 4.25a.75.75 0 0 1-1.06-1.06L11.94 10L8.22 6.28a.75.75 0 0 1 0-1.06" clip-rule="evenodd"/></svg> <div class="sui-accordion-summary-content">  </div> </button> <div class="sui-accordion-details" aria-hidden="true"> <div class="sui-accordion-content"> Content for item 2 </div> </div> </div>
<div class="sui-accordion-item" data-open="false"> <button class="sui-accordion-summary not-content" type="button" aria-expanded="false"> <svg style="min-width: 20px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="sui-summary-chevron" height="20" width="20" viewbox="0 0 20 20"><path fill="currentColor" fill-rule="evenodd" d="M8.22 5.22a.75.75 0 0 1 1.06 0l4.25 4.25a.75.75 0 0 1 0 1.06l-4.25 4.25a.75.75 0 0 1-1.06-1.06L11.94 10L8.22 6.28a.75.75 0 0 1 0-1.06" clip-rule="evenodd"/></svg> <div class="sui-accordion-summary-content">  </div> </button> <div class="sui-accordion-details" aria-hidden="true"> <div class="sui-accordion-content"> Content for item 3 </div> </div> </div> </div> <script type="module" src="/mock/path/Accordion.astro?astro&type=script&index=0&lang.ts"></script>"
`;

exports[`Accordion Component > renders AccordionItem component correctly 1`] = `"<div class="sui-accordion-item" data-open="false"> <button class="sui-accordion-summary not-content" type="button" aria-expanded="false"> <svg style="min-width: 20px" xmlns="http://www.w3.org/2000/svg" prefix="heroicons" class="sui-summary-chevron" height="20" width="20" viewbox="0 0 20 20"><path fill="currentColor" fill-rule="evenodd" d="M8.22 5.22a.75.75 0 0 1 1.06 0l4.25 4.25a.75.75 0 0 1 0 1.06l-4.25 4.25a.75.75 0 0 1-1.06-1.06L11.94 10L8.22 6.28a.75.75 0 0 1 0-1.06" clip-rule="evenodd"/></svg> <div class="sui-accordion-summary-content">  </div> </button> <div class="sui-accordion-details" aria-hidden="true"> <div class="sui-accordion-content">  </div> </div> </div>"`;
